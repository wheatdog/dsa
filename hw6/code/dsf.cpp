#include <vector>
#include "dsf.h"
extern "C" 
{
#include "avl_ntudsa.h"
}
#ifdef DEBUG
#include <iostream>
#endif

int Compare_by_int(const void *pa, const void *pb, void *param)
{
    int a = *(const int *)pa;
    int b = *(const int *)pb;

    if (a > b) return 1;
    else if (a < b) return -1;
    else return 0;
}

DSF::DSF(int num, std::vector<int> priceTable)
{
    for (int count = 0; count <= num; count++) {
        elements.push_back(count);
        ownerTable.push_back(count);
        depthTable.push_back(1);

        struct avl_table *tmp;
        tmp = avl_create(Compare_by_int, NULL, NULL);
        avl_probe(tmp, priceTable[count]);

        price_tree_table.push_back(tmp);
    }
}

int DSF::Find(int n)
{
    while (elements[n] != n)
        n = elements[n];
    return n;
}

void insert(struct avl_table *target, struct avl_node *node)
{
    if (node == NULL)
        return;

    for (int count = 0; count < node->avl_cnt; count++) {
        avl_probe(target, node->avl_data);
    }


    for (int branch = 0; branch < 2; branch++)
        insert(target, node->avl_link[branch]);
}

void DSF::Union(int first, int second)
{
    ownerTable[second] = Owner(first);
    ownerTable[first] = Owner(first);

    if (depthTable[second] < depthTable[first]) {
        elements[second] = first;

        struct avl_table *tmp =  price_tree_table[second];
        struct avl_table *target =  price_tree_table[first];
        insert(target, tmp->avl_root);
    }
    else if (depthTable[second] > depthTable[first]) {
        elements[first] = second;

        struct avl_table *tmp =  price_tree_table[first];
        struct avl_table *target =  price_tree_table[second];
        insert(target, tmp->avl_root);
    }
    else {
        elements[second] = first;
        depthTable[first]++;

        struct avl_table *tmp =  price_tree_table[second];
        struct avl_table *target =  price_tree_table[first];
        insert(target, tmp->avl_root);
    }

#ifdef DEBUG
    for (unsigned int i = 153; i < 160; i++)
        std::cout <<  elements[i] << " ";
    std::cout << std::endl;

    for (unsigned int i = 153; i < 160; i++)
        std::cout <<  depthTable[i] << " ";
    std::cout << std::endl;

    for (unsigned int i = 153; i < 160; i++)
        std::cout <<  ownerTable[i] << " ";
    std::cout << std::endl;
#endif 
}

int DSF::Owner(int n)
{
    return ownerTable[n];
}

struct avl_table* DSF::GetTree(int n)
{
    return price_tree_table[n];
}
