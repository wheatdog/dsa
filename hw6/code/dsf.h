#ifndef DSF_H
#define DSF_H 1

#include "avl_ntudsa.h"

#endif /* dsf.h */
