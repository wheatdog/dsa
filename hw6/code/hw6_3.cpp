#include <iostream>
#include "dsf.h"

using namespace std;

int main()
{
    int friendNum, incidentNum;
    cin >> friendNum >> incidentNum;

    int price[friendNum];
    for (int count = 0; count < friendNum; count++)
    {
        cin >> price[count];
    }

    int type;
    while (cin >> type)
    {
        switch(type)
        {
            case 1:
            {
                int tediousID, interestID;
                cin >> tediousID >> interestID;
            } break;

            case 2:
            {
                int gameID;
                long long int money;
                cin >> gameID >> money;
            } break;
        }
    }

    return 0;
}

