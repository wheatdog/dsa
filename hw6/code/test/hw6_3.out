3 0
3 1
1 2
1 2
4 1
1 3
